static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,340 @@
-                <allow>
-                    <!-- allow dependency within the package -->
-                    <from class="teammates.test.pageobjects.*" />
-                </allow>
-                <allow>
-                    <!-- allow dependency from UI tests -->
-                    <from class="teammates.test.cases.browsertests.*" />
-                </allow>
-            </deny>
-        </access-rule>
+    
+    <ruleset name="Architectural rules for test code">
+        <ruleset name="TestDriverComponent: Test cases should not be dependent on each other">
+            <var name="testCasesPackage" value="teammates.test.cases" />
+            <access-rule>
+                <deny>
+                    <from class="${testCasesPackage}.**" />
+                    <to class="${testCasesPackage}.**" />
+                    <allow>
+                        <!-- allow dependency from inner class -->
+                        <from class="${testCasesPackage}.**$**" />
+                    </allow>
+                    <allow>
+                        <to>
+                            <!-- allow dependency to inner class -->
+                            <include class="${testCasesPackage}.**$**" />
+                            <!-- allow dependency to base class -->
+                            <include class="${testCasesPackage}.**Base**" />
+                        </to>
+                    </allow>
+                    <allow>
+                        <!-- allow dependency from Feedback*QuestionUiTest to FeedbackQuestionUiTest -->
+                        <from class="${testCasesPackage}.browsertests.Feedback*QuestionUiTest" />
+                        <to class="${testCasesPackage}.browsertests.FeedbackQuestionUiTest" />
+                    </allow>
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="TestDriverComponent: Only UI tests can access page object classes">
+            <access-rule>
+                <deny>
+                    <from class="teammates.test.**" />
+                    <to class="teammates.test.pageobjects.**" />
+                    <allow>
+                        <from>
+                            <include class="teammates.test.pageobjects.**" />
+                            <include class="teammates.test.cases.browsertests.**" />
+                        </from>
+                    </allow>
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="TestDriverComponent: Test driver can only be accessed from test package">
+            <access-rule>
+                <deny>
+                    <from class="teammates.**" />
+                    <to class="teammates.test.driver.**" />
+                    <allow>
+                        <from>
+                            <include class="teammates.test.**" />
+                            <include class="teammates.client.**" />
```

Comment:
The way this rule is configured, it is actually not adding any enforcement. If a production code tries to use test/client code, a compile error is guaranteed.